### PR TITLE
[Misc] Remove local DeepSeek V2 model fork

### PR DESCRIPTION
**Pull request by YunzhuLu**
## What this RP do

This PR removes the local DeepSeek V2/V3.2 model fork (`deepseek_v2.py` / `deepseek_mtp.py`) and switches those architectures back to upstream vLLM.

Kunlun-specific Indexer and Sparse MLA behavior is moved out of the model file into OOT ops, an Indexer backend/builder.

### Goal

Delete the model definitions in `deepseek_v2.py` and stop overriding `DeepseekV3ForCausalLM` / `DeepseekV32ForCausalLM` / `GlmMoeDsaForCausalLM` / `DeepSeekMTP` in the Kunlun model registry.

### Implementation

- Register `KunlunSparseAttnIndexer` as the OOT `SparseAttnIndexer` (old `vllm::sparse_attn_indexer_vllm_kunlun` path is not kept).
- Add `KunlunDeepseekV32IndexerBackend` + metadata builder for Kunlun Indexer fields (`context_*_lens`, decode `seq_lens_cpu`).
- Patch only `DeepseekV32IndexerCache.get_attn_backend()` to return the Kunlun backend.
- Use upstream `Attention`; set `FlashMLASparseBackend.get_name()` to `FLASHMLA_SPARSE`.
- Drop the local DeepSeek / MTP / GLM DSA model registry overrides.
- New Features from upstream such as DCP, compressed indexer cache, and FP4 are explicitly unsupported.

### Three remaining patches
All in `_apply_indexer_patch`, for the reasons below :
- `DeepseekV32IndexerCache.get_attn_backend()` is a hardcoded `return DeepseekV32IndexerBackend` and never consults the platform, so there is no hook to register a Kunlun backend against.
- `per_token_group_quant_fp8` -> `indexer_quant2d`: P800 has no FP8 compute, and the quant call sits before the OOT boundary, so by the time `forward_oot` runs `q_scale` is already folded into `weights` and unrecoverable.
- `_build_prefill_chunk_metadata_kernel` -> torch: `super().build()` launches it unconditionally, and the community build without Triton.

### Known limitations
- Fail-fast with `NotImplementedError`: `num_speculative_tokens >= 2`  (upstream's flattening layout), DCP, compressed indexer cache, FP4 cache. 
- `pack_seq_triton` / `unpack_seq_triton` are still called on the decode padding edge path (full-CGs decode and num_reqs < capture size). Same as the old fork, but still Triton;
- Collides with #439: both register OOT under the same key and `register_oot` asserts uniqueness, so whichever lands second fails at import.
- **No hardware validation yet**; a P800 smoke test for prefill/decode/MTP is still needed.

### 1. Q quantization
Upstream `Indexer.forward` emits FP8 (`fused_indexer_q_rope_quant`) and folds `q_scale * softmax_scale * head_scale` into `weights`. Kunlun `I8_mqa_logits` / `I8_paged_mqa_logits` still expect int8 Q from Indexer-side `quant2d`, with `scale / 127` folded into `weights`. This is a contract mismatch at the Indexer, not inside OOT. By the time `forward_oot` runs, `q_scale` is gone. Bitcast (`.view(int8)`) or requantizing FP8 in OOT double-counts or misreads the scale. This PR
fails fast on non-int8.
Three ways to restore a real run (this PR does none of them):
- **A. Keep int8, no model fork.** Patch upstream `Indexer.forward` the same way this PR patches `get_attn_backend()`: skip `fused_indexer_q_rope_quant`, do RoPE then `quant2d`, fold `scale / 127`. Do not to keep another `deepseek_v2.py`; only the Q path changes. Current `I8_*` kernels stay. This is the intended follow-up if we keep today's kernels.
- **B. Keep int8 by keeping part of the model.** Leave a local `Indexer`  (or a model override) that still does `quant2d`. That is a thinner fork, not a full `deepseek_v2.py`, but it is still a model-side split. This PR is trying to avoid that.
- **C. Follow upstream Q, new kernels.** Leave Indexer on FP8 and replace `I8_*` with FP8 MQA logits ops. No model fork, no Indexer quant patch; **requires new Kunlun operators**.

### 2. Triton
Need a confirmation first: can this Kunlun image actually launch upstream Triton kernels? The old fork still calls `pack_seq_triton` on an edge path, but #404 says Kunlun cannot JIT/launch Triton (`CUDA_ERROR_NOT_SUPPORTED`). The Indexer Builder should follow that answer: keep `super().build()` only if those kernels run; otherwise fill the same fields without launching them.

- *New in this PR — `super().build()`.* The Kunlun Builder now calls upstream `DeepseekV32IndexerMetadataBuilder.build()`, which can launch Triton metadata kernels (`_build_prefill_chunk_metadata_kernel`, and on some decode paths `_prepare_uniform_decode_kernel`). The old fork did not go through this builder. If the current Kunlun image cannot JIT/launch those kernels, metadata construction fails before Indexer runtime. This needs a hardware check. If it cannot run, the Builder must fill the same fields without that launch.
- *Old path — `pack_seq_triton` / `unpack_seq_triton`.* The old fork already used these for `requires_padding` (uneven decode / short chunked prefill). They are not new and are not on the common equal-length prefill/decode path. This PR keeps them for parity. Need to confirm on hardware later; failure there does not block deleting the model.

### Follow-ups

Need to merge this `SparseAttnIndexer` OOT with #439 into a single class. Both use the same registration key; a second OOT register will fail.

### Validation

- [x] `pre-commit run --all-files`
- [x] Confirm DeepSeek V3.2 loads the upstream model (no local `deepseek_v2.py` override).
- [x] Confirm `DeepseekV32IndexerCache.get_attn_backend()` returns `KunlunDeepseekV32IndexerBackend`.
- [ ] Confirm `SparseAttnIndexer` constructs `KunlunSparseAttnIndexer`.
- [ ] Dummy / first forward: non-int8 Q raises `NotImplementedError` (expected until Q quant is decided).
- [ ] After Q path is fixed: smoke prefill + decode for DeepSeek V3.2.
- [ ] Optional: equal-length MTP decode; DCP/FP4/compress still fail fast.

<details>
<summary><b>Detailed Checklist (Click to Expand)</b></summary>

<p>Thank you for contributing to <b>vLLM Kunlun</b>!
To help us maintain high code quality and streamline the review process, please ensure your PR meets the following requirements.</p>

<h3>1. Code Quality</h3>

<ul>
    <li>All linting and formatting checks pass (<code>pre-commit</code>).</li>
    <li>The code is well-structured and sufficiently documented.</li>
    <li>The change is designed with maintainability and readability in mind.</li>
</ul>

<h3>2. Testing</h3>

<ul>
    <li>Relevant unit tests are added or updated.</li>
    <li>Integration tests are included when applicable.</li>
    <li>Existing tests continue to pass.</li>
</ul>

<h3>3. DCO Compliance</h3>

<p>This project follows the
<a href="https://github.com/vllm-project/vllm/blob/main/DCO">Developer Certificate of Origin (DCO)</a>.</p>

<ul>
    <li>All commits include a <code>Signed-off-by:</code> line.</li>
    <li>Use <code>git commit -s</code> to automatically add the sign-off.</li>
</ul>

<h3>4. Review Expectations</h3>

<p>During the review process, maintainers may:</p>

<ul>
    <li>Request code refactoring or additional tests.</li>
    <li>Ask for clarifications on design decisions.</li>
    <li>Suggest performance, stability, or maintainability improvements.</li>
</ul>

<p>We appreciate your patience and collaboration throughout the review process!</p>

</details>

### CODE REVIEW

**Comment by YunzhuLu** [comment]
Hi @GrootLiu , this PR removes the local DeepSeek fork. The change isn't small — could you take a look at the design and the blockers?

File: vllm_kunlun/v1/attention/backends/mla/flashmla_sparse.py
Comment (GrootLiu) [inline vllm_kunlun/v1/attention/backends/mla/flashmla_sparse.py:1]: 这是为了对齐上游的枚举是吗

**Comment by GrootLiu** [comment]
> ## What this RP do
> This PR removes the local DeepSeek V2/V3.2 model fork (`deepseek_v2.py` / `deepseek_mtp.py`) and switches those architectures back to upstream vLLM.
> 
> Kunlun-specific Indexer and Sparse MLA behavior is moved out of the model file into OOT ops, an Indexer backend/builder
> 
> ### Goal
> Delete the model definitions in `deepseek_v2.py` and stop overriding `DeepseekV3ForCausalLM` / `DeepseekV32ForCausalLM` / `GlmMoeDsaForCausalLM` / `DeepSeekMTP` in the Kunlun model registry.
> 
> ### Implementation
> * Register `KunlunSparseAttnIndexer` as the OOT `SparseAttnIndexer` (old `vllm::sparse_attn_indexer_vllm_kunlun` path is not kept).
> * Add `KunlunDeepseekV32IndexerBackend` + metadata builder for Kunlun Indexer fields (`context_*_lens`, decode `seq_lens_cpu`).
> * Patch only `DeepseekV32IndexerCache.get_attn_backend()` to return the Kunlun backend.
> * Use upstream `Attention`; set `FlashMLASparseBackend.get_name()` to `FLASHMLA_SPARSE`.
> * Drop the local DeepSeek / MTP / GLM DSA model registry overrides.
> * New Features from upstream such as DCP, compressed indexer cache, and FP4 are explicitly unsupported.
> 
> ## Blockers
> ### 1. Q quantization
> Upstream `Indexer.forward` emits FP8 (`fused_indexer_q_rope_quant`) and folds `q_scale * softmax_scale * head_scale` into `weights`. Kunlun `I8_mqa_logits` / `I8_paged_mqa_logits` still expect int8 Q from Indexer-side `quant2d`, with `scale / 127` folded into `weights`. This is a contract mismatch at the Indexer, not inside OOT. By the time `forward_oot` runs, `q_scale` is gone. Bitcast (`.view(int8)`) or requantizing FP8 in OOT double-counts or misreads the scale. This PR fails fast on non-int8. Three ways to restore a real run (this PR does none of them):
> 
> * **A. Keep int8, no model fork.** Patch upstream `Indexer.forward` the same way this PR patches `get_attn_backend()`: skip `fused_indexer_q_rope_quant`, do RoPE then `quant2d`, fold `scale / 127`. Do not to keep another `deepseek_v2.py`; only the Q path changes. Current `I8_*` kernels stay. This is the intended follow-up if we keep today's kernels.
> * **B. Keep int8 by keeping part of the model.** Leave a local `Indexer`  (or a model override) that still does `quant2d`. That is a thinner fork, not a full `deepseek_v2.py`, but it is still a model-side split. This PR is trying to avoid that.
> * **C. Follow upstream Q, new kernels.** Leave Indexer on FP8 and replace `I8_*` with FP8 MQA logits ops. No model fork, no Indexer quant patch; **requires new Kunlun operators**.
> 
> ### 2. Triton
> Need a confirmation first: can this Kunlun image actually launch upstream Triton kernels? The old fork still calls `pack_seq_triton` on an edge path, but #404 says Kunlun cannot JIT/launch Triton (`CUDA_ERROR_NOT_SUPPORTED`). The Indexer Builder should follow that answer: keep `super().build()` only if those kernels run; otherwise fill the same fields without launching them.
> 
> * _New in this PR — `super().build()`._ The Kunlun Builder now calls upstream `DeepseekV32IndexerMetadataBuilder.build()`, which can launch Triton metadata kernels (`_build_prefill_chunk_metadata_kernel`, and on some decode paths `_prepare_uniform_decode_kernel`). The old fork did not go through this builder. If the current Kunlun image cannot JIT/launch those kernels, metadata construction fails before Indexer runtime. This needs a hardware check. If it cannot run, the Builder must fill the same fields without that launch.
> * _Old path — `pack_seq_triton` / `unpack_seq_triton`._ The old fork already used these for `requires_padding` (uneven decode / short chunked prefill). They are not new and are not on the common equal-length prefill/decode path. This PR keeps them for parity. Need to confirm on hardware later; failure there does not block deleting the model.
> 
> ### Follow-ups
> Need to merge this `SparseAttnIndexer` OOT with #439 into a single class. Both use the same registration key; a second OOT register will fail.
> 
> ### Validation
> * [ ]  `pre-commit run --all-files`
> * [ ]  Confirm DeepSeek V3.2 loads the upstream model (no local `deepseek_v2.py` override).
> * [ ]  Confirm `DeepseekV32IndexerCache.get_attn_backend()` returns `KunlunDeepseekV32IndexerBackend`.
> * [ ]  Confirm `SparseAttnIndexer` constructs `KunlunSparseAttnIndexer`.
> * [ ]  Dummy / first forward: non-int8 Q raises `NotImplementedError` (expected until Q quant is decided).
> * [ ]  After Q path is fixed: smoke prefill + decode for DeepSeek V3.2.
> * [ ]  Optional: equal-length MTP decode; DCP/FP4/compress still fail fast.
> 
> Detailed Checklist (Click to Expand)

对于第一个问题来说，目前第一种解决方案比较好，第二种不具备长期可维护性，也不符合我们一开始的目标，第三种解决方案由于 P800 不具备 FP8 计算能力，应该不会开发对应的 Kernel

再回答第二个问题，虽然有针对 Kunlun 开发的 xTriton，但是其性能在 P800 上不太理想，因此我们在社区中不使用任何 Triton 能力

**Comment by GrootLiu** [comment]
@feiyu-byte 也可以同步看看这个 PR

File: vllm_kunlun/v1/attention/backends/mla/flashmla_sparse.py
Comment (YunzhuLu) [reply vllm_kunlun/v1/attention/backends/mla/flashmla_sparse.py:1]: 是的，虽然 Kunlun 不依靠名称作为枚举key创建 AttentionBackend，但 `MLAAttention` 初始化时会根据 `attn_backend.get_name()` ，将 fp8 等量化 KV cache dtype 转化为 'fp8_ds_mla', 如果继续返回`FLASHMLA_SPARSE_VLLM_V1`，用户初始化时指定的 fp8 和其他类型名称不会被正确转化，可能导致 Kunlun 的 `FlashMLASparseImpl.forward()` 没办法拦截住当前不支持的量化类型。

**Comment by YunzhuLu** [comment]
Hi @GrootLiu, the two blockers from review are resolved
- Q quant: option A (`per_token_group_quant_fp8` -> `indexer_quant2d`)
- No Triton in `super().build()`: torch metadata kernel + fail-fast on DCP / compress / flattening

Thanks for taking another look

File: vllm_kunlun/ops/sparse_attn_indexer.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+from vllm.v1.attention.ops.common import pack_seq_triton, unpack_seq_triton
+
+from vllm_kunlun.ops.deep_gemm import int8_mqa_logits, int8_paged_mqa_logits
+from vllm_kunlun.v1.attention.backends.mla.indexer import (
+    KunlunDeepSeekV32IndexerDecodeMetadata,
+    KunlunDeepseekV32IndexerPrefillChunkMetadata,
+)
+
+
+def indexer_quant2d(
```
Comment (GrootLiu) [inline vllm_kunlun/ops/sparse_attn_indexer.py:15]: 这里目前使用 pytorch 写的算子，后续可以对 xspeedgate 团队提算子需求～

**Review by GrootLiu — APPROVED** [review]
LGTM
